File: iosApp/iosApp/Forum/ForumViewModel.swift
import Foundation
import Combine
import Shared

/// ViewModel for the Forum tab following the MVI pattern.
@MainActor
final class ForumViewModel: ObservableObject {

    @Published private(set) var state = ForumState()

    private let effectSubject = PassthroughSubject<ForumEffect, Never>()
    var effectPublisher: AnyPublisher<ForumEffect, Never> {
        effectSubject.eraseToAnyPublisher()
    }

    private let getThreads: GetThreadsUseCase = RaceDependencyProvider.companion.shared.createGetThreadsUseCase()

    init() {
        loadThreads()
    }

    func send(_ intent: ForumIntent) {
        switch intent {
        case .refresh:
            loadThreads()
        case .dismissError:
            state.errorMessage = nil
        case .selectSort(let sort):
            state.selectedSort = sort
            loadThreads()
        case .selectCategory(let category):
            state.selectedCategory = category
            loadThreads()
        }
    }

    func loadThreads() {
        Task { await performLoad() }
    }

    func refresh() async {
        await performLoad()
    }

    private func performLoad() async {
        state.isLoading = true
        state.errorMessage = nil

        do {
            let result = try await getThreads.invoke(
                sort: state.selectedSort,
                category: state.selectedCategory,
                userId: nil
            )
            state.isLoading = false
            // On failure, keep the threads already on screen and show why.
            if let threads = result.data as? [Shared.Thread] {
                state.threads = threads
            }
            state.errorMessage = result.error?.message
        } catch {
            state.isLoading = false
            state.errorMessage = error.localizedDescription
        }
    }
}

File: composeApp/src/androidMain/kotlin/org/gce/racehub/forum/ForumViewModel.kt
package org.gce.racehub.forum

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.receiveAsFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import org.gce.racehub.auth.domain.session.UserSession
import org.gce.racehub.race.domain.usecase.GetThreadsUseCase

class ForumViewModel(
    private val getThreadsUseCase: GetThreadsUseCase,
    private val userSession: UserSession
) : ViewModel() {

    private val _state = MutableStateFlow(ForumState())
    val state: StateFlow<ForumState> = _state.asStateFlow()

    private val _effect = Channel<ForumEffect>(Channel.BUFFERED)
    val effect = _effect.receiveAsFlow()

    init {
        loadThreads()
    }

    fun onIntent(intent: ForumIntent) {
        when (intent) {
            is ForumIntent.Refresh -> loadThreads()
            is ForumIntent.DismissError -> _state.update { it.copy(errorMessage = null) }
            is ForumIntent.SelectSort -> {
                _state.update { it.copy(selectedSort = intent.sort) }
                loadThreads()
            }
            is ForumIntent.SelectCategory -> {
                _state.update { it.copy(selectedCategory = intent.category) }
                loadThreads()
            }
        }
    }

    private fun loadThreads() {
        viewModelScope.launch {
            _state.update { it.copy(isLoading = true, errorMessage = null) }
            val result = getThreadsUseCase(
                sort = _state.value.selectedSort,
                category = _state.value.selectedCategory,
                userId = userSession.userId
            )
            // On failure, keep the threads already on screen and show why the refresh failed.
            _state.update {
                it.copy(
                    isLoading = false,
                    threads = result.data ?: it.threads,
                    errorMessage = result.error?.message
                )
            }
        }
    }
}
